### PR TITLE
"(Deprecated)" should show in version select list in installed tab right pane

### PR DESCRIPTION
**Pull request by scottbommarito**
## Bug

Fixes: part of https://github.com/NuGet/Home/issues/8516 and https://github.com/NuGet/Home/issues/8541
Regression: No
* Last working version:   n/a
* How are we preventing it in future:   n/a

## Fix

Changed the version list show "(Deprecated)" alongside deprecated versions.

image

Unfortunately, this only applies to the Installed tab for now as the Browse tab doesn't have the deprecation metadata available when it renders the list.

This will make it easier for users to determine which version they should choose and prevent them from updating to deprecated versions.

## Testing/Validation

Tests Added: No
Reason for not adding tests:  UI stuff
Validation:  worked on my machine for packages with/without deprecated versions in both project and solution view

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Globalization;
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:5]: > Unfortunately, this only applies to the Installed tab for now as the Browse tab doesn't have the deprecation metadata available when it renders the list.

Can you elaborate on why this won't work with browse?

Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:5]: The list of versions that appears in this list is fetched from search in the Browse tab (which doesn't have deprecation metadata) and from registration in the Installed tab (which does), regardless of the fact that it ultimately fetches the full metadata from registration in both tabs.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs [hunk 2]
```diff
@@ -48,22 +51,28 @@ public class DisplayVersion
...
+            IsDeprecated = isDeprecated;
 
             // Display a single version if the range is locked
-            if (range.HasLowerAndUpperBounds && range.MinVersion == range.MaxVersion)
-            {
-                var formattedVersionString = Version.ToString(versionFormat, VersionFormatter.Instance);
+            // If range is unlocked, display it and use the original value for floating ranges
+            var versionString = range.HasLowerAndUpperBounds && range.MinVersion == range.MaxVersion ?
+                Version.ToString(versionFormat, VersionFormatter.Instance)
+                : Range.OriginalString;
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs]: Personally, I prefer the format
```csharp
var variable = expr
  ? value1
  : value2;
```
but right now the `?` and `:` positions are inconsistent.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs [hunk 3]
```diff
@@ -48,22 +51,28 @@ public class DisplayVersion
...
+
+            _toString += versionString;
+
+            if (IsDeprecated)
             {
-                // Display the range, use the original value for floating ranges
-                _toString = string.IsNullOrEmpty(AdditionalInfo) ?
-                    Range.OriginalString :
-                    AdditionalInfo + " " + Range.OriginalString;
+                _toString += string.Format(
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:75]: Avoid refactoring all of the above, if you're only gonna append Deprecated to the string. 

We need to mindful of unnecessary changes this late in the game.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs [hunk 4]
```diff
@@ -48,22 +51,28 @@ public class DisplayVersion
...
+                    CultureInfo.CurrentCulture,
+                    "    ({0})",
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:77]: Feels like there's more than one space. How many should there be?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs [hunk 5]
```diff
@@ -48,22 +51,28 @@ public class DisplayVersion
...
-                    formattedVersionString :
-                    AdditionalInfo + " " + formattedVersionString;
+            _toString = "";
+            if (!string.IsNullOrEmpty(AdditionalInfo))
+            {
+                _toString += AdditionalInfo + " ";
             }
-            else
+
+            _toString += versionString;
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs]: I'm not a fan of multiple string concatenations. Ok, there's only between 1 and 4 concatenations depending on the code path taken, and with this few string concat usually has better perf than creating a new string builder and using it. But I'm sure there can be better performing code while still being easy to read.

How about something like:

```csharp
string formatString;
if (!string.IsNullOrEmpty(AdditionalInfo))
{
  formatString = IsDeprecated
    ? "{0} {1}   ({2})"
    : "{0} {1}";
}
else
{
  formatString = IsDeprecated
    ? "{1}   ({2})"
    : "{1}";
}

_toString = string.Format(formatString, AdditionalInfo, versionString, Resources.Label_Deprecated);
```

although given there's only 4 formats, maybe don't use `formatString`, and just assign to `_toString` in a single `String.Concat`.

**Review by nkolev92 — COMMENTED** [review]
Looks good overall. 
Would like to see some tests if possible. 

Ideally a vendor run as well.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 6]
```diff
@@ -143,7 +143,11 @@ public virtual void CleanUp()
             }
 
             // Add Current package version to package versions list.
-            _allPackageVersions = new List<NuGetVersion>() { searchResultPackage.Version };
+            _allPackageVersions = new List<(NuGetVersion version, bool isDeprecated)>()
+            {
+                (searchResultPackage.Version, (await searchResultPackage.GetPackageDeprecationMetadataAsync()) != null)
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: We've already discussed this as too much to refactor, but the lack of separation between UI and business logic code makes me sad. Particularly since I've been working on some UI delay issues.

Can `searchResultPackage.GetPackageDeprecationMetadataAsync()` make an HTTP call, or is the data already in memory, and it's just searching data structures? Does this code happen on the UI thread, or a background thread?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs [hunk 7]
```diff
@@ -157,9 +161,9 @@ public virtual void CleanUp()
             }
 
             // Get the list of available versions, ignoring null versions
-            _allPackageVersions = versions
+            _allPackageVersions = (await Task.WhenAll(versions
                 .Where(v => v?.Version != null)
-                .Select(v => v.Version)
+                .Select(async v => (v.Version, (await (v.PackageSearchMetadata?.GetDeprecationMetadataAsync() ?? Task.FromResult((PackageDeprecationMetadata)null))) != null))))
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: can this be simplified? perhaps by creating a method?  `Task.FromResult` is usually in a method that returns a `Task` but is not async, but this lamba method is async. And the `!= null` at the end  is confusing. Plus the `ToList` on the line below isn't running on the same sequence that `Where` and `Select` are, so it would be good if they were indented differently to make it more clear. The code before was easy to understand, this new code is not.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 4 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:77]: I had one space at first, and then @rrelyea suggested adding multiple spaces and we thought it looked better. There's 4 I think? Unfortunately we're heavily restricted with how the text is formatted and displayed due to the nature of ComboBox and felt this was the easiest stopgap.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 5 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs]: Since those strings are localized, consider creating a string resource for each case. That makes the code easier to debug and it also makes easier to avoid spelling errors on the UI.

Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs]: The format strings are not localised, and the other strings are already localised. I don't understand which strings you're proposing should be in the resources.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs (hunk 6 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: I completely agree. I think this code will need a large refactor eventually or else the situation will continue to get progressively worse, as it was fundamentally written in a way that did not separate the UI layer from the business logic.

I looked into this more. It is wrong to fetch the deprecation information at this point--the idea is that the code inserts the current version with nothing else to bypass the case in which fetching the versions (or, in this case, the deprecation metadata) takes a long time. I should insert "false" here instead of making a call.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 5 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs]: I agree, but due to @nkolev92's concerns I'm going to simplify the amount of changes here to avoid push-back.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs (hunk 7 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Models/DetailControlModel.cs]: created a method as requested.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:75]: good call

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 2 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs]: no longer an issue since refactoring removed

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 3 above)
Comment (scottbommarito) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:75]: note that the refactoring above was removed but this comment was on a line that didn't change

File: src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs (hunk 4 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/DisplayVersion.cs:77]: Best effort :)

**Review by nkolev92 — APPROVED** [review]
